pkg/manager/member/scaler.go

```diff
@@ -104,3 +113,91 @@ func ordinalPVCName(memberType v1alpha1.MemberType, setName string, ordinal int3
 func ordinalPodName(memberType v1alpha1.MemberType, tcName string, ordinal int32) string {
 	return fmt.Sprintf("%s-%s-%d", tcName, memberType, ordinal)
 }
+
+func validateScaling(actual *apps.StatefulSet, desired *apps.StatefulSet) error {
+	if !features.DefaultFeatureGate.Enabled(features.AdvancedStatefulSet) {
+		return nil
+	}
+	// Following cases are not supported yet.
+	// TODO support it or validate in validation phase
+	actualDeleteSlots := helper.GetDeleteSlots(actual)
+	desiredDeleteSlots := helper.GetDeleteSlots(desired)
+	if *actual.Spec.Replicas == *desired.Spec.Replicas {
+		// no scaling
+		if actualDeleteSlots.Equal(desiredDeleteSlots) {
+			return nil
+		}
+	} else if *actual.Spec.Replicas > *desired.Spec.Replicas {
+		// scaling in
+		if actualDeleteSlots.Equal(desiredDeleteSlots) || desiredDeleteSlots.IsSuperset(actualDeleteSlots) {
+			return nil
+		}
+	} else {
+		// scaling out
+		if actualDeleteSlots.Equal(desiredDeleteSlots) || actualDeleteSlots.IsSuperset(desiredDeleteSlots) {
+			return nil
+		}
+	}
+	return fmt.Errorf("unsupported scaling operation for sts %s/%s (actual: %d, %v; desired: %d, %v)", actual.Namespace, actual.Name,
+		*actual.Spec.Replicas, actualDeleteSlots.List(),
+		*desired.Spec.Replicas, desiredDeleteSlots.List(),
+	)
+}
+
+// return the last ordinal in the list, -1 if the list is empty
+func lastOrdinalInList(list []int) int {
+	if len(list) == 0 {
+		return -1
+	}
+	return list[len(list)-1]
+}
+
+// scaleOne scales actual set to desired set by deleting or creating a replica
+func scaleOne(actual *apps.StatefulSet, desired *apps.StatefulSet) (int32, int32, sets.Int, error) {
```

Comment:
this function calculates desired `replicas` and `delete slots` from actual/desired stateful sets by allowing one pod to be deleted or created